File: nodes/array-unique/array-unique.process.py
import json

async def process(inputs, settings, config, nodeConfig):
  array = inputs.get("array", [])
  if not isinstance(array, list):
    array = [array]
    
  mode = settings.get("mode", "value")
  property_name = settings.get("property")
  case_sensitive = settings.get("case_sensitive", True)
  
  def get_compare_value(item):
    value = item
    
    if mode == "property" and property_name:
      if isinstance(item, dict):
        value = item.get(property_name)
      else:
        value = None
        
    if mode == "json":
      try:
        return json.dumps(item, sort_keys=True)
      except:
        return str(item)
        
    if isinstance(value, str) and not case_sensitive:
      return value.lower()
      
    return value
    
  seen = {}
  duplicates = []
  indices = []
  
  for i, item in enumerate(array):
    compare_value = get_compare_value(item)
    # Use repr for non-string keys to handle unhashable types
    key = compare_value if isinstance(compare_value, (str, int, float, bool)) else repr(compare_value)
    
    if key not in seen:
      seen[key] = (item, i)
    else:
      duplicates.append(item)
      
  unique = []
  for item, index in seen.values():
    unique.append(item)
    indices.append(index)
    
  return {
    "array": unique,
    "count": len(unique),
    "duplicates": duplicates,
    "indices": indices
  } 

File: nodes/array-unique/array-unique.process.js
export default async ({inputs, settings, config}) => {

  const array = Array.isArray(inputs.array) ? inputs.array : [inputs.array];
  const mode = settings.mode || "value";
  const property = settings.property;
  const caseSensitive = settings.case_sensitive !== false;

  const getCompareValue = (item) => {
    let value = item;
    
    if (mode === "property" && property && typeof item === 'object' && item !== null) {
      value = item[property];
    }
    
    if (mode === "json") {
      try {
        // Sort keys to ensure consistent ordering for object comparison
        return JSON.stringify(item, Object.keys(item).sort());
      } catch {
        return String(item);
      }
    }
    
    if (typeof value === 'string' && !caseSensitive) {
      return value.toLowerCase();
    }
    
    return value;
  };

  const seen = new Map();
  const duplicates = [];
  const indices = [];
  
  array.forEach((item, index) => {
    const compareValue = getCompareValue(item);
    if (!seen.has(compareValue)) {
      seen.set(compareValue, {item, index});
    } else {
      duplicates.push(item);
    }
  });

  const unique = Array.from(seen.values()).map(({item, index}) => {
    indices.push(index);
    return item;
  });

  return {
    array: unique,
    count: unique.length,
    duplicates: duplicates,
    indices: indices
  };
}; 